src/mapping/query/function/parse_url.rs

```diff
@@ -0,0 +1,129 @@
+use super::prelude::*;
+use std::collections::BTreeMap;
+use std::iter::FromIterator;
+use url::Url;
+
+#[derive(Debug)]
+pub(in crate::mapping) struct ParseUrlFn {
+    query: Box<dyn Function>,
+}
+
+impl ParseUrlFn {
+    #[cfg(test)]
+    pub(in crate::mapping) fn new(query: Box<dyn Function>) -> Self {
+        Self { query }
+    }
+}
+
+impl Function for ParseUrlFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let bytes = required!(ctx, self.query, Value::Bytes(v) => v);
+
+        Url::parse(&String::from_utf8_lossy(&bytes))
+            .map_err(|e| format!("unable to parse url: {}", e))
+            .map(Into::into)
+    }
+
+    fn parameters() -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            accepts: |v| matches!(v, Value::Bytes(_)),
+            required: true,
+        }]
+    }
+}
+
+impl TryFrom<ArgumentList> for ParseUrlFn {
+    type Error = String;
+
+    fn try_from(mut arguments: ArgumentList) -> Result<Self> {
+        let query = arguments.required("value")?;
+
+        Ok(Self { query })
+    }
+}
+
+impl From<Url> for Value {
+    fn from(url: Url) -> Self {
+        let mut map = BTreeMap::<&str, Value>::new();
+
+        map.insert("scheme", url.scheme().to_owned().into());
+        map.insert("username", url.username().to_owned().into());
+        map.insert("path", url.path().to_owned().into());
+        map.insert("password", url.password().map(ToOwned::to_owned).into());
+        map.insert("host", url.host_str().map(ToOwned::to_owned).into());
+        map.insert("port", url.port().map(|v| v as isize).into());
+        map.insert("fragment", url.fragment().map(ToOwned::to_owned).into());
+        map.insert(
+            "query",
+            url.query_pairs()
+                .into_owned()
+                .map(|(k, v)| (k, v.into()))
+                .collect::<BTreeMap<String, Value>>()
+                .into(),
+        );
```

Comment:
I chose to keep all fields, even if they contain no data (`null` or empty map `{}`). This improves discoverability for the end-user to know which properties can be manipulated.

The (to be introduced) `format_url` will ignore `null` values as expected.

Reply:
I also think it makes sense to introduce a `compact` function eventually, to allow removing empty data (`null`, `[]`, `{}` or even empty strings (all configurable)) from events.

Reply:
Created #4430 to track compacting events.